File: RenderServer/src/main/java/me/hydos/alchemytools/network/netty/MessageEncoder.java
package me.hydos.alchemytools.network.netty;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToByteEncoder;

public class MessageEncoder extends MessageToByteEncoder<ChannelMessage> {

    @Override
    protected void encode(ChannelHandlerContext ctx, ChannelMessage msg, ByteBuf out) {
        out.writeShort(msg.packetId());
        out.writeByte(msg.direction());
        out.writeBytes(msg.packetData());
    }
}

File: RenderClient/netty/MessageEncoder.cs
using DotNetty.Buffers;
using DotNetty.Codecs;
using DotNetty.Transport.Channels;

namespace RenderClient.netty; 

public class MessageEncoder : MessageToByteEncoder<ChannelMessage> {

    protected override void Encode(IChannelHandlerContext context, ChannelMessage msg, IByteBuffer output) {
        output.WriteShort(msg.PacketId);
        output.WriteByte(msg.Direction);
        output.WriteInt(msg.PacketData.Length);
        output.WriteBytes(msg.PacketData);
    }
}